src/validation/code.rs

```diff
@@ -357,7 +357,120 @@ fn read_instructions(
     }
 }
 
-fn validate_value_stack<F>(return_ty: ResultType, f: F) -> Result<()>
+pub fn read_constant_instructions(
+    wasm: &mut WasmReader,
+    value_stack: &mut Vec<ValType>,
+    _globals_ty: &[GlobalType],
+) -> Result<Span> {
+    let start_pc = wasm.pc;
+
+    let assert_pop_value_stack = |value_stack: &mut Vec<ValType>, expected_ty: ValType| {
+        value_stack
+            .pop()
+            .ok_or(Error::InvalidValueStackType(None))
+            .and_then(|ty| {
+                (ty == expected_ty)
+                    .then_some(())
+                    .ok_or(Error::InvalidValueStackType(Some(ty)))
+            })
+    };
+
+    loop {
+        let Ok(first_instr_byte) = wasm.read_u8() else {
+            return Err(Error::ExprMissingEnd);
+        };
+        trace!("Read cosntant instruction byte {first_instr_byte:#X?} ({first_instr_byte})");
+
+        // Valid constant instructions are:
+        // - Core: https://webassembly.github.io/spec/core/valid/instructions.html#valid-constant
+        // - Extended Proposal: https://webassembly.github.io/extended-const/core/valid/instructions.html#valid-constant
+        use crate::core::reader::types::opcode::*;
+        match first_instr_byte {
+            // Missing: ref.null, ref.func, global.get
```

Comment:
These extensive comments are really nice, can we make them doc string of the surrounding function?